# Regular Expressions
## Trevor, Kate, Abe

Import packages

In [1]:
import re
import pandas as pd
from collections import Counter

In [2]:
with open(r'web_log_file.txt', 'r') as file:
    webLog =  file.read()
    
# It works but it's really long so don't try to print it out

In [3]:
webLog[0:2000]

'\n109.169.248.247 - - [12/Dec/2015:18:25:11 +0100] "GET /administrator/ HTTP/1.1" 200 4263 "-" "Mozilla/5.0 (Windows NT 6.0; rv:34.0) Gecko/20100101 Firefox/34.0" "-"\n109.169.248.247 - - [12/Dec/2015:18:25:11 +0100] "POST /administrator/index.php HTTP/1.1" 200 4494 "http://almhuette-raith.at/administrator/" "Mozilla/5.0 (Windows NT 6.0; rv:34.0) Gecko/20100101 Firefox/34.0" "-"\n46.72.177.4 - - [12/Dec/2015:18:31:08 +0100] "GET /administrator/ HTTP/1.1" 200 4263 "-" "Mozilla/5.0 (Windows NT 6.0; rv:34.0) Gecko/20100101 Firefox/34.0" "-"\n46.72.177.4 - - [12/Dec/2015:18:31:08 +0100] "POST /administrator/index.php HTTP/1.1" 200 4494 "http://almhuette-raith.at/administrator/" "Mozilla/5.0 (Windows NT 6.0; rv:34.0) Gecko/20100101 Firefox/34.0" "-"\n83.167.113.100 - - [12/Dec/2015:18:31:25 +0100] "GET /administrator/ HTTP/1.1" 200 4263 "-" "Mozilla/5.0 (Windows NT 6.0; rv:34.0) Gecko/20100101 Firefox/34.0" "-"\n83.167.113.100 - - [12/Dec/2015:18:31:25 +0100] "POST /administrator/index.php

Splitting each line by IP address to better look at each entry, then printing the total number of entries

In [4]:
splitLog = webLog.split('\n')

In [5]:
print(splitLog[1], '\n')
print(splitLog[2], '\n')
print(splitLog[3], '\n')

len(splitLog)

109.169.248.247 - - [12/Dec/2015:18:25:11 +0100] "GET /administrator/ HTTP/1.1" 200 4263 "-" "Mozilla/5.0 (Windows NT 6.0; rv:34.0) Gecko/20100101 Firefox/34.0" "-" 

109.169.248.247 - - [12/Dec/2015:18:25:11 +0100] "POST /administrator/index.php HTTP/1.1" 200 4494 "http://almhuette-raith.at/administrator/" "Mozilla/5.0 (Windows NT 6.0; rv:34.0) Gecko/20100101 Firefox/34.0" "-" 

46.72.177.4 - - [12/Dec/2015:18:31:08 +0100] "GET /administrator/ HTTP/1.1" 200 4263 "-" "Mozilla/5.0 (Windows NT 6.0; rv:34.0) Gecko/20100101 Firefox/34.0" "-" 



22141

# 1. What browsers are used to access this website? Are there some that are more common than others? (Best guesses on browsers is good enough here, some are difficult to identify)

From looking at the individual logs from above, we see that the browser is always listed at the end of the entry in between the characters "-" "-". Below I create a regular expression to take the values between those characters, create a dataframe with those strings, and see the individual value counts of each. 

In [6]:
pattern_1 = '(?<="-" ")(.*)(?=" "-")'
Browsers = re.findall(pattern_1, webLog)
Browsers
BROWSERS = pd.DataFrame(Browsers,columns=['Browsers'])
BROWSERS["Browsers"].value_counts()

unique = pd.DataFrame(BROWSERS["Browsers"].value_counts())

unique.head(10)



Browsers
Mozilla/5.0 (Windows NT 6.0; rv:34.0) Gecko/201...      2572
-                                                       2088
Python-urllib/2.7                                        457
Mozilla/5.0 (compatible; bingbot/2.0; +http://w...       403
Python-urllib/1.17                                       368
Mozilla/5.0 (Windows NT 5.1; rv:29.0) Gecko/201...       354
Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.1 ...       256
Mozilla/5.0 (compatible; Googlebot/2.1; +http:/...       218
Mozilla/5.0 (compatible; Baiduspider/2.0; +http...       216
Mozilla/5.0 (compatible; YandexBot/3.0; +http:/...        85

After taking a look at some of the top values, I ran into some interesting things:


1) the second most frequent value is just a hyphen - I looked further into this and found those entries looked something like this:  (208.43.103.162 - - [16/Dec/2015:17:01:31 +0100] "GET / HTTP/1.1" 200 10439 "-" "-" "-"), I don't necessarily know what that means, but it doesn't look like your average browser access.



2) The second and third most used browsers were something called "Python-urllib" and "bingbot.htm" from Mozilla. Both are also not your average browsers (may be some indication of malicious entry attempts)



3) The part that I really didn't understand was trying to distinguish between google chrome and safari - it seems that every entry mentioning google chrome also mentioned safari - because of this I queried all of the times Chrome was mentioned, and subtracted that from the total mentions of safari in order to get some more accurate numbers. (I also referenced this website to get a better understanding of the connection labels : https://www.whatismybrowser.com/guides/the-latest-user-agent/chrome)





In [7]:
#printing out the value counts for all browser labels
unique.head(10)

Browsers
Mozilla/5.0 (Windows NT 6.0; rv:34.0) Gecko/201...      2572
-                                                       2088
Python-urllib/2.7                                        457
Mozilla/5.0 (compatible; bingbot/2.0; +http://w...       403
Python-urllib/1.17                                       368
Mozilla/5.0 (Windows NT 5.1; rv:29.0) Gecko/201...       354
Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.1 ...       256
Mozilla/5.0 (compatible; Googlebot/2.1; +http:/...       218
Mozilla/5.0 (compatible; Baiduspider/2.0; +http...       216
Mozilla/5.0 (compatible; YandexBot/3.0; +http:/...        85

After debating what to put for the final answer, I chose to stick with the main browsers that I could easily identify: Chrome, Safari and Firefox. The sell above can be easily modified to remove the head function so one could easily go identify other browsers used, (but I believe these 3 to be the most frequently used). One last thing to mention, there also appear to be alot of connection attempts with different "bots" - whether this is malicious or not will need further attention, but I felt it important to note. 

In [8]:
pattern_1 = 'Chrome'
len(re.findall(pattern_1, webLog))

4553

In [9]:
pattern_2 = 'Safari'
len(re.findall(pattern_2, webLog))

7272

In [10]:
pattern_3 = 'Firefox'
len(re.findall(pattern_3, webLog))

7715

From the above code, we can see that Chrome was used to access 4553 times, Safari was used 7272 - 5443 = 1829 times, and Firefox was used 7715 times. Combining this with the over 2000 entries labled "-" and nearly 1000 more from the python package, we can account for a large ammount of the 22,000 total entries. 

Of the most common consumer used browsers, Firefox was used most often, followed by Chrome and then Safari.

# 2. What websites/IP addresses bring this website the most traffic?

Again from looking at the individual entries, we can see that IP addresses are the first thing listed per connection. Below is a regular expression to grab the values between the new line (indicating a new entry) and "- -" (what looks like the indication that the IP address stops. I then print out all of the IP addresses queried. 

In [11]:
pattern_1 = '(?<=\n)(.*)(?= - -)'
IP_Adresses = re.findall(pattern_1, webLog)
IP = pd.DataFrame(IP_Adresses,columns=['IP'])
IP.head()

IP
0  109.169.248.247
1  109.169.248.247
2      46.72.177.4
3      46.72.177.4
4   83.167.113.100

I then put these values into a dataframe, and take the 5 most commonly occuring IP addresses to see which brought the most traffic to the website. 

In [12]:
IP = pd.DataFrame(IP_Adresses,columns=['IP'])
IP["IP"].value_counts().head(5)

148.251.50.49     1929
52.22.118.215      734
213.150.254.81     434
37.1.206.196       433
84.112.161.41      376
Name: IP, dtype: int64

On top of the IP addresses listed above, we can look back to the code from the browser searches and see the following websites create significant traffic (or are atleast mentioned in the connection):

http://www.bing.com/bingbot.htm

http://www.google.com/bot.html

http://www.baidu.com/search/spider.html

http://yandex.com/bots



# 3. Are there specific times of the year that the website recieves more traffic than other times?

In [13]:
from datetime import datetime

In [14]:
#Getting all dates from the log
pattern = '\d+/\w+/\d+'
dates = list()
for line in splitLog:
    dates.append(re.findall(pattern, line))

In [15]:
OurDates = pd.DataFrame(dates, columns = ["Date"])
OurDates = OurDates[1:]
OurDates['Date'] = OurDates['Date'].apply(lambda x: datetime.strptime(x, '%d/%b/%Y'))
OurDates['Day of the Week'] = OurDates['Date'].apply(lambda x: x.weekday())
OurDates['Month'] = OurDates['Date'].apply(lambda x: x.month) 
OurDates['Week'] = OurDates['Date'].apply(lambda x: x.isocalendar()[1])
OurDates

Date  Day of the Week  Month  Week
1     2015-12-12                5     12    50
2     2015-12-12                5     12    50
3     2015-12-12                5     12    50
4     2015-12-12                5     12    50
5     2015-12-12                5     12    50
...          ...              ...    ...   ...
22136 2016-01-16                5      1     2
22137 2016-01-16                5      1     2
22138 2016-01-16                5      1     2
22139 2016-01-16                5      1     2
22140 2016-01-16                5      1     2

[22140 rows x 4 columns]

In [44]:
#print(OurDates.value_counts().head(10))

print(OurDates["Day of the Week"].value_counts().head(10))
print(OurDates["Week"].value_counts().head(10))
print(OurDates["Month"].value_counts().head(10))

OurDates[OurDates['Week'] == 53]


5    3943
0    3417
6    3332
2    3137
3    3108
4    2758
1    2445
Name: Day of the Week, dtype: int64
51    7904
2     3873
53    3202
1     2947
52    2495
50    1719
Name: Week, dtype: int64
12    14148
1      7992
Name: Month, dtype: int64


Date  Day of the Week  Month  Week
12119 2015-12-28                0     12    53
12120 2015-12-28                0     12    53
12121 2015-12-28                0     12    53
12122 2015-12-28                0     12    53
12123 2015-12-28                0     12    53
...          ...              ...    ...   ...
15316 2016-01-03                6      1    53
15317 2016-01-03                6      1    53
15318 2016-01-03                6      1    53
15319 2016-01-03                6      1    53
15320 2016-01-03                6      1    53

[3202 rows x 4 columns]

Looking at the dates from the log, it appears that we only have dates from December 2015 through January 2016. However, from these two months we can see that some days had more traffic than others. The top dates for traffic were December 19th, 13th, and 18th in 2015.

If we look at each day of the week specifically, it appears that Saturdays had the most traffic and Tuesdays had the least.

While the logs don't include the full months of December and January, we can see that from what we do have that December had about twice the traffic of January.

As far as which week was the busiest of the logs that we have, we can see that the 51st week of the year was very busy which was 12-14-2015 through 12-20-2015. It had over twice as much traffic as the other weeks.

# 4.  This log file is supposedly contains some malicious attempts to alter the site. Do you see any evidence of this? 

In [ ]:
#log file for reference and a starting glance
#for line in splitLog:
    #print(line)

One place there might be some malicious attempts to the website are where there are many requests from the same IP or lots of failed logins. Additionaly if there is a new spike that could be a signal. I looked up and saw that there were lots of entires that look similar to

52.22.118.215 - - [11/Jan/2016:11:43:25 +0100] "GET /apache-log/access_150111.log HTTP/1.0" 404 222 "-" "Python-urllib/1.17" "-"

From what I understand It looks like someone is trying to get access to the log but with the 404 it does not look like they are sucessful. So I'm going to explore this a little more. 

This is just setting a pattern to look for all of the the apache-log/access attempts. It seemed best to just search by this becuase it would get the most hits in the easiet way. 

In [ ]:
pattern = '(apache-log\/access_)'
loc = re.search(pattern, webLog)
print(loc)

In [ ]:
print(webLog[int(loc.start()):int(loc.end())])


Function to look through the entire log file line by line

In [ ]:
def match_reg(pattern, string):
    return(re.match(pattern, string))
  
pattern = r'(.*:).*(apache-log\/access_)'

matches = list()
for line in splitLog:
    
    m = match_reg(pattern, line)
    
    if m != None:
        matches.append((m.group(1).replace(':', ''), m.group(2)))

Counting all the instances where there is the pattern of apache-log/acess_ ...

In [ ]:
apache_log = [x[0] for x in matches]

count_title_mentions = Counter(apache_log)

In [ ]:
print(count_title_mentions)

I don't think it's helpful to put this on a bar chart becuase clearly this IP 52.22.118.215 has 100's of attempts within seconds to get the log that they are being lumped together. In total 52.22.118.215 has 732 (255+208+158+111) on January 11th for the same thing. 

Adidtionally the IP 76.168.128.240 has a total of 87 (46+32+9) attemps on the same day. I would say this looks pretty suspicious  as there are no other similar atempts and so many on the same day.

I am going to do the same process again but this time with looking for sql.

In [ ]:
pattern = '(sql)'
loc = re.search(pattern, webLog)
print(loc)

print(webLog[int(loc.start()):int(loc.end())])


In [ ]:
def match_reg(pattern, string):
    return(re.match(pattern, string))
  
pattern = r'(.*:).*(sql)'

matches = list()
for line in splitLog:
    
    m = match_reg(pattern, line)
    
    if m != None:
        matches.append((m.group(1).replace(':', ''), m.group(2)))

In [ ]:
sql_log = [x[0] for x in matches]

count_sql_mentions = Counter(sql_log)

Honestly, I don't really konw what this means or if there is valuable information so I won't try to break it down too much but from intuition it looks like there might be some sql attemps to break in becuase I read that sometimes people use things that look like a sql attempt to get in. It looks like that could be the case here but I don't know enought about log files to see if it's useful. 

In [ ]:
print(count_sql_mentions)